planner/core/preprocess.go

```diff
@@ -1430,15 +1449,27 @@ func (p *preprocessor) stmtType() string {
 
 func (p *preprocessor) handleTableName(tn *ast.TableName) {
 	if tn.Schema.L == "" {
-		if _, ok := p.withName[tn.Name.L]; ok {
-			return
-		}
-
 		currentDB := p.ctx.GetSessionVars().CurrentDB
 		if currentDB == "" {
 			p.err = errors.Trace(ErrNoDB)
 			return
 		}
+
+		if p.preprocessWith.containsCTEName(tn.Name.L) {
+			if !p.preprocessWith.inWithClause {
+				return
+			}
+
+			if !p.ensureInfoSchema().TableExists(model.NewCIStr(currentDB), tn.Name) {
+				return
+			}
```

Comment:
Why do we need to check this?

Reply:
For some case.If the innermost `CTE name` is the same as `table name` and  the CTE is non recursive, we should add schema to the `table name`(t1).
```
create view v1 as with t1 as (
  select a from t2 where t2.a=3
  union
  select t2.a+1 from t1,t2 where t1.a=t2.a)
select * from t1;
```